### PR TITLE
feat(clr): Add Implementation for hipInitDevice API

### PR DESCRIPTION
**Pull request by satyanveshd**
## Motivation

Adds implementation for missing API hipInitDevice in HIP runtime

## Technical Details

hipInitDevice initializes the resources for the specified device and sets the passed device flags along with initializing the default stream on that device.

## Issue Tracking

JIRA ID:  AIRUNTIME-2600

## Test Plan

Add catch tests with few positive and negative scenarios. Also add contract tests for the new API.

## Test Result

All the added tests passed with the changes in the PR built locally.

## Submission Checklist

- [x] Look over the contributing guidelines at https://github.com/ROCm/rocm-systems/blob/develop/CONTRIBUTING.md.

### CODE REVIEW

File: projects/clr/hipamd/include/hip/amd_detail/hip_api_trace.hpp [hunk 1]
```diff
@@ -48,7 +48,7 @@
 #define HIP_API_TABLE_STEP_VERSION 0
 #define HIP_COMPILER_API_TABLE_STEP_VERSION 0
 #define HIP_TOOLS_API_TABLE_STEP_VERSION 1
-#define HIP_RUNTIME_API_TABLE_STEP_VERSION 31
+#define HIP_RUNTIME_API_TABLE_STEP_VERSION 32
```
Comment (jaydeeppatel1111) [inline projects/clr/hipamd/include/hip/amd_detail/hip_api_trace.hpp]: Bumping `HIP_RUNTIME_API_TABLE_STEP_VERSION` to 32 may breaks the rocprofiler-sdk build. Need abi adjustment in rocprofiler side.

File: projects/hipother/hipnv/include/hip/nvidia_detail/nvidia_hip_runtime_api.h [hunk 2]
```diff
@@ -2045,6 +2045,10 @@ inline static hipError_t hipInit(unsigned int flags) {
   return hipCUResultTohipError(cuInit(flags));
 }
 
+inline static hipError_t hipInitDevice(int device, unsigned int deviceFlags, unsigned int flags) {
```
Comment (jaydeeppatel1111) [inline projects/hipother/hipnv/include/hip/nvidia_detail/nvidia_hip_runtime_api.h:2052]: may need CUDA version guard?

File: projects/hip/include/hip/hip_runtime_api.h [hunk 3]
```diff
@@ -1014,6 +1014,8 @@ enum hipLimit_t {
 #define hipDeviceScheduleMask 0x7
 #define hipDeviceMapHost 0x8
 #define hipDeviceLmemResizeToMax 0x10
+/** Indicates that the deviceFlags passed to hipInitDevice() are valid and should be applied.*/
+#define hipInitDeviceFlagsAreValid 0x1
```
Comment (jaydeeppatel1111) [inline projects/hip/include/hip/hip_runtime_api.h:1019]: Nvidia mapping?

File: projects/clr/hipamd/src/hip_device_runtime.cpp [hunk 4]
```diff
@@ -840,7 +840,47 @@ hipError_t hipSetDeviceFlags(unsigned int flags) {
...
+  // Apply the device scheduling flags only when hipInitDeviceFlagsAreValid is passed.
+  if (flags == hipInitDeviceFlagsAreValid) {
+    hipError_t err = ihipSetDeviceFlags(dev, deviceFlags);
+    if (err != hipSuccess) {
+      HIP_RETURN(err);
+    }
+  }
+
+  // creates default (null) stream on the initialized device.
+  dev->NullStream();
```
[the remark this answers is not in the document]
Comment (satyanveshd) [reply projects/clr/hipamd/src/hip_device_runtime.cpp:889]: Done.

File: projects/clr/hipamd/include/hip/amd_detail/hip_api_trace.hpp (hunk 1 above)
Comment (satyanveshd) [reply projects/clr/hipamd/include/hip/amd_detail/hip_api_trace.hpp]: The step version checks are already added in rocprofiler-sdk.

File: projects/hip/include/hip/hip_runtime_api.h (hunk 3 above)
Comment (satyanveshd) [reply projects/hip/include/hip/hip_runtime_api.h:1019]: Done.

File: projects/hipother/hipnv/include/hip/nvidia_detail/nvidia_hip_runtime_api.h (hunk 2 above)
Comment (satyanveshd) [reply projects/hipother/hipnv/include/hip/nvidia_detail/nvidia_hip_runtime_api.h:2052]: Done.

**Review by gandryey — APPROVED** [review]

**Comment by satyanveshd** [comment]
Can someone from @ROCm/rocm-devtools-team @ROCm/rocprofiler-owners give approval? Thanks.

**Review by bgopesh — COMMENTED** [review]
Changes look good from rocprofiler-sdk side. Waiting for a rebase to see the sdk tests passing.

**Review by bgopesh — APPROVED** [review]

**Comment by litvaOo** [comment]
Only rocgdb-cpu is failing, all approvals and required jobs are present. Overriding
